### PR TITLE
feat(object): allow publish once to all enviroments 

### PR DESCRIPTION
**Pull request by JacopoDaeli**
Additionally provides an endpoint to fetch all heads (head in each environment).

### CODE REVIEW

**Comment by JacopoDaeli** [comment]
PS: I can split the PR into two smaller PRs if it makes review easier.

File: lib/plugins/object.js [hunk 1]
```diff
@@ -434,22 +437,27 @@ module.exports = fp(
...
+      async function putObjectVariant(
+        {
+          name: keyname,
+          version,
+          env,
+          variant = '_default',
+          expiration = null,
+          data
+        },
+        { forceCreateEnv = false } = { forceCreateEnv: false }
```
Comment (jpage-godaddy) [inline lib/plugins/object.js:459]: Minor, but could be:

```suggestion
        { forceCreateEnv = false } = { }
```

File: lib/routes/objects.js [hunk 2]
```diff
@@ -144,6 +144,54 @@ module.exports =
...
+            name: { type: 'string' }
+          }
+        },
+        response: {
+          200: {
+            type: 'array',
+            items: {
+              type: 'object',
+              properties: {
+                enviroment: { type: 'string' },
```
Comment (jpage-godaddy) [inline lib/routes/objects.js:163]: Should this be `environment`?

Comment (jpage-godaddy) [inline lib/routes/objects.js:163]:
```suggestion
                environment: { type: 'string' },
```

File: lib/routes/objects.js [hunk 3]
```diff
@@ -144,6 +144,54 @@ module.exports =
...
+          const {
+            params: { name }
+          } = req;
+          const envs = await fastify.getEnvs({ name });
+          const heads = await Promise.all(
+            envs.map(async ({ env }) => {
+              const obj = await fastify.getObject({ name, env });
+              if (!obj) return null;
+              const { headVersion, latestVersion } = obj;
+              return { enviroment: env, headVersion, latestVersion };
```
Comment (jpage-godaddy) [inline lib/routes/objects.js:186]:
```suggestion
              return { environment: env, headVersion, latestVersion };
```

File: lib/routes/objects.js [hunk 4]
```diff
@@ -220,41 +268,98 @@ module.exports =
...
+                category: 'database', // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html#ecs-event-category-database
+                // eslint-disable-next-line max-len
+                type: ['creation', 'allowed'], // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-creation
+                method: req.method,
+                url: req.url,
+                requestId: req.id,
+                sourceAddress: req.ip,
+                host: req.hostname
+              });
+            }
```
Comment (jpage-godaddy) [inline lib/routes/objects.js:323]:
```suggestion
              fastify.log.security?.({
                success: true,
                message: `Object "${name}" variant "${variant}" sucessfully created in "${env}"`,
                // eslint-disable-next-line max-len
                category: 'database', // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html#ecs-event-category-database
                // eslint-disable-next-line max-len
                type: ['creation', 'allowed'], // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-creation
                method: req.method,
                url: req.url,
                requestId: req.id,
                sourceAddress: req.ip,
                host: req.hostname
              });
```

File: lib/routes/objects.js [hunk 5]
```diff
@@ -220,41 +268,98 @@ module.exports =
...
```
Comment (jpage-godaddy) [inline lib/routes/objects.js:362]:
```suggestion
              fastify.log.security?.({
                success: true,
                message: `Object "${name}" variant "${variant}" sucessfully created in "${envs.join(
                  ', '
                )}"`,
                // eslint-disable-next-line max-len
                category: 'database', // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html#ecs-event-category-database
                // eslint-disable-next-line max-len
                type: ['creation', 'allowed'], // see https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-type.html#ecs-event-type-creation
                method: req.method,
                url: req.url,
                requestId: req.id,
                sourceAddress: req.ip,
                host: req.hostname
              });
```

File: test/routes/envs.test.js [hunk 6]
```diff
@@ -71,10 +65,10 @@ test('Enviroments API', async (t) => {
       })
     });
 
-    t.equal(res2.statusCode, 409);
+    t.equal(res.statusCode, 409);
 
-    const body2 = JSON.parse(res2.payload);
-    t.same(body2.message, 'Enviroment already exists');
+    const body = JSON.parse(res.payload);
+    t.same(body.message, 'Enviroment already exists');
```
Comment (jpage-godaddy) [inline test/routes/envs.test.js:71]:
```suggestion
    t.same(body.message, 'Environment already exists');
```

File: test/routes/objects.test.js [hunk 7]
```diff
@@ -536,7 +575,154 @@ test('Objects API', async (t) => {
...
+
+    const body = JSON.parse(res.payload);
+    t.same(body, { created: true });
+
+    const head = await getHeads(fastify, {
+      name: 'preObj1'
+    });
+    t.same(head, [
+      {
+        enviroment: 'prod',
```
Comment (jpage-godaddy) [inline test/routes/objects.test.js:623]: Basically find/replace all "enviroment" with "environment".

```suggestion
        environment: 'prod',
```

File: lib/routes/objects.js (hunk 3 above)
Comment (akerr-godaddy) [reply lib/routes/objects.js:186]: I see a bunch more instances of this. @JacopoDaeli you'll want to do a find/replace all

**Comment by JacopoDaeli** [comment]
@dcousineau @dcousineau-godaddy feedbacks?

File: lib/routes/objects.js (hunk 4 above)
Comment (JacopoDaeli) [reply lib/routes/objects.js:323]: I will change all the occurrences like you suggested with a sperate PR. Out of scope in this one.

File: lib/routes/objects.js (hunk 5 above)
Comment (JacopoDaeli) [reply lib/routes/objects.js:362]: Same here.

**Comment by JacopoDaeli** [comment]
@jpage-godaddy ready for a new review.

**Review by jpage-godaddy — APPROVED** [review]

**Pull request merged**